indexing-service/src/main/java/io/druid/indexing/common/task/ParallelIndexSubTask.java

```diff
@@ -0,0 +1,451 @@
+            fireDepartmentMetrics.incrementThrownAway();
+            continue;
+          }
+
+          if (!Intervals.ETERNITY.contains(inputRow.getTimestamp())) {
+            final String errorMsg = StringUtils.format(
+                "Encountered row with timestamp that cannot be represented as a long: [%s]",
+                inputRow
+            );
+            throw new ParseException(errorMsg);
+          }
+
+          if (explicitIntervals) {
+            final Optional<Interval> optInterval = granularitySpec.bucketInterval(inputRow.getTimestamp());
+            if (!optInterval.isPresent()) {
+              fireDepartmentMetrics.incrementThrownAway();
+              continue;
+            }
+          }
+
+          // Segments are created as needed, using a single sequence name. They may be allocated from the overlord
+          // (in append mode) or may be created on our own authority (in overwrite mode).
+          final String sequenceName = getId();
+          final AppenderatorDriverAddResult addResult = driver.add(inputRow, sequenceName);
+
+          if (addResult.isOk()) {
+            if (exceedMaxRowsInSegment(addResult.getNumRowsInSegment(), tuningConfig) ||
+                exceedMaxRowsInAppenderator(addResult.getTotalNumRowsInAppenderator(), tuningConfig)) {
+              // There can be some segments waiting for being published even though any rows won't be added to them.
+              // If those segments are not published here, the available space in appenderator will be kept to be small
+              // which makes the size of segments smaller.
+              final SegmentsAndMetadata pushed = driver.pushAllAndClear(pushTimeout);
+              pushedSegments.addAll(pushed.getSegments());
+              log.info("Pushed segments[%s]", pushed.getSegments());
+            }
+          } else {
+            throw new ISE("Failed to add a row with timestamp[%s]", inputRow.getTimestamp());
+          }
+
+          fireDepartmentMetrics.incrementProcessed();
+        }
+        catch (ParseException e) {
+          if (tuningConfig.isReportParseExceptions()) {
+            throw e;
+          } else {
+            fireDepartmentMetrics.incrementUnparseable();
+          }
+        }
+      }
+
+      final SegmentsAndMetadata pushed = driver.pushAllAndClear(pushTimeout);
+      pushedSegments.addAll(pushed.getSegments());
+      log.info("Pushed segments[%s]", pushed.getSegments());
+
+      return pushedSegments;
+    }
+    catch (TimeoutException | ExecutionException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  private static boolean exceedMaxRowsInSegment(
+      int numRowsInSegment,
+      ParallelIndexTuningConfig indexTuningConfig
+  )
+  {
+    // maxRowsInSegment should be null if numShards is set in indexTuningConfig
+    final Integer maxRowsInSegment = indexTuningConfig.getTargetPartitionSize();
+    return maxRowsInSegment != null && maxRowsInSegment <= numRowsInSegment;
```

Comment:
This means `max` is really `flush boundary` not `max`, is that clarified in the docs?

Reply:
What docs are you mean? `targetPartitionSize` is already well explained, I think (http://druid.io/docs/latest/ingestion/tasks.html#tuningconfig).